base/unchecked.jl

```diff
@@ -0,0 +1,266 @@
+# working fine in LLVM. (Note that division is still handled via `base/int.jl`,
+# which always checks for overflow, and which provides its own sets of
+# work-arounds for LLVM codegen bugs.) However, the comments in `base/int.jl`
+# and in issue #4905 are more pessimistic. For the time being, we thus retain
+# the ability to handle codegen bugs in LLVM, until the code here has been
+# tested on more systems and architectures. It also seems that things depend on
+# which compiler that was used to build LLVM (i.e. either gcc or clang).
+
+# These unions are used for most unchecked functions:
+#     BrokenSignedInt
+#     BrokenUnsignedInt
+# These unions are used for unchecked_{mul,div,rem}:
+#     BrokenSignedIntMul
+#     BrokenUnsignedIntMul
+
+typealias BrokenSignedInt Union{}
+typealias BrokenUnsignedInt Union{}
+if WORD_SIZE == 32
+    typealias BrokenSignedIntMul Int128
+    typealias BrokenUnsignedIntMul UInt128
+else
+    typealias BrokenSignedIntMul Union{}
+    typealias BrokenUnsignedIntMul Union{}
+end
+# Use these definitions to test the non-LLVM implementations
+# typealias BrokenSignedInt SignedInt
+# typealias BrokenUnsignedInt UnsignedInt
+# typealias BrokenSignedIntMul SignedInt
+# typealias BrokenUnsignedIntMul UnsignedInt
+
+"""
+    Base.unchecked_neg(x)
+
+Calculates `-x` without any overflow checking. It is the caller's
+responsiblity to ensure that there is no overflow, and the compiler is free to
+optimize the code assuming there is no overflow.
+"""
+function unchecked_neg end
+
+function unchecked_neg{T<:SignedInt}(x::T)
+    box(T, unchecked_sneg_int(unbox(T,x)))
+end
+function unchecked_neg{T<:UnsignedInt}(x::T)
+    box(T, unchecked_uneg_int(unbox(T,x)))
+end
+unchecked_neg{T<:BrokenSignedInt}(x::T) = -x
+unchecked_neg{T<:BrokenUnsignedInt}(x::T) = T(0)
+
+"""
+    Base.unchecked_abs(x)
+
+Calculates `abs(x)` without any overflow checking. It is the caller's
+responsiblity to ensure that there is no overflow, and the compiler is free to
+optimize the code assuming there is no overflow.
+"""
+function unchecked_abs end
+
+unchecked_abs(x::SignedInt) = ifelse(x<0, -x, x)
+unchecked_abs(x::UnsignedInt) = x
+
+"""
+    Base.unchecked_add(x, y)
+
+Calculates `x+y` without any overflow checking. It is the caller's
+responsiblity to ensure that there is no overflow, and the compiler is free to
+optimize the code assuming there is no overflow.
+"""
+function unchecked_add end
+
+unchecked_add(x::Integer, y::Integer) = unchecked_add(promote(x,y)...)
+unchecked_add{T<:Integer}(x::T, y::T) = x+y
+function unchecked_add{T<:SignedInt}(x::T, y::T)
+    box(T, unchecked_sadd_int(unbox(T,x), unbox(T,y)))
+end
+function unchecked_add{T<:UnsignedInt}(x::T, y::T)
+    box(T, unchecked_uadd_int(unbox(T,x), unbox(T,y)))
+end
+unchecked_add{T<:BrokenSignedInt}(x::T, y::T) = x+y
+unchecked_add{T<:BrokenUnsignedInt}(x::T, y::T) = x+y
+
+# Handle multiple arguments
+unchecked_add(x) = x
+unchecked_add{T}(x1::T, x2::T, x3::T) =
```

Comment:
You should be able to define
```
@inline unchecked_add(x1, x2, x...) = unchecked_add(unchecked_add(x1, x2), x...)
```
The `@inline` means it will be expanded at compile-time. Whether you're using generated functions or this more "lispy" construct, you'll fall catastrophically over a cliff beyond 8 arguments.